### PR TITLE
[ BB2-964 ] Full Release Process Orchestration via Jenkins

### PR DESCRIPTION
**Pull request by oragame**
**JIRA Ticket:**
[BB2-964](https://jira.cms.gov/browse/BB2-964)

**User Story or Bug Summary:**

Automate the build, deploy and migrations for a release. Migrations can be run pre-deploy or post deploy or not at all depending on the needs.

### What Does This PR Do?
New Jenkinsfile created for automation to streamline the deployment process.

**Required parameters**
* App Environment (choice)
* App Migrations (choice)
* App repo branch (string)
* Deploy repo branch (string)

**Optional Parameters**
* Platinum AMI ID (use most current if blank)
* App AMI ID (If provided, build stage is skipped. If blank, build stage will create the AMI ID)

The existing BUILD - App AMI (AMZN2) and DEPLOY - App AMI jobs are chained to this automation. This will allow for manual process to remain, and prevent duplication (for now). If at a future time the BUILD and DEPLOY jobs need to be customized for either manual or auto deploys, those steps can be added to this automation to allow for that.

### What Should Reviewers Watch For?

For testing, only the TEST environment option is available in the drop down until thoroughly tested. 'Z' Jenkins Jobs are used in the test process to prevent unnecessary job runs on the production BUILD and DEPLOY jobs.

* Is this likely to address the goals expressed in the user story?
* Are any additional documentation updates needed?
* Are there any unhandled and/or untested edge cases you can think of?
* Is user input properly sanitized & handled?
* Does this make any backwards-incompatible changes that might break end user clients?
* Can you find any bugs if you run the code locally and test it manually?

If you're reviewing this PR, please check these things, in particular:

* Open to suggestions for job name, currently it is: Z AUTO - Build Deploy Migrate App AMI
* For Post-deployment migrations, I have left the input option to Migrate/Abort - My gut feeling is to leave the option for user decision in case a rollback is needed and discovered prior to choosing 'Migrate'. Any thoughts around this are welcome.
* If an APP_AMI_ID is provided, the build stage will be skipped. Most of the time this value will not be provided, but it is available in the event that a STABLE BUILD has already been run for this release.
* Are there any optimizations that may have been overlooked? As far as efficiency of chaining the jobs etc..

**Test Scenarios:**
* Test each migration option
* Test blank vs provided Platinum AMI's and App AMI's (make sure build stage is skipped if App AMI provided)
* Once all scenarios pass on TEST, determine best time to test on IMPL and PROD

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then here's a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence: N/A.
    * Does this PR add any new software dependencies? **No**.
    * Does this PR modify or invalidate any of our security controls? **No**.
    * Does this PR store or transmit data that was not stored or transmitted before? **No**.
* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR should not be merged unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons? **No**.

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following pre-requisite changes have been fully deployed:

* N/A

### Submitter Checklist

I have gone through and verified that...:

* [X] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [X] I have named this PR and its branch such that they'll be automatically be linked to the (most) relevant Jira issue, per: <https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html>.
* [X] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [X] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [X] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [X] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. BFD, SLS, etc.).
* [X] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

**Comment by oragame** [comment]
Testing update: 
* Build/Deploy with no migrations is testing OK
* Pre-deploy migrations failing on 'migrate' command with `pgaudit must be loaded via shared_preload_libraries`

I'll research how to work around this error, and I think once that is fixed the complete automation flow should work with pre or post migrations.

**Comment by oragame** [comment]
**Testing update:**
After Nathan's modification to the DB, I was able to perform the complete process with build / migrate and deploy. This is all working well.

**Examples jobs can be reviewed here:**
* [Build / Pre-deploy migration / Deploy](https://ci.backends.cms.gov/non-oc/job/Blue%20Button/job/Z%20AUTO%20-%20Build%20Deploy%20Migrate%20App%20AMI/13/console) (TEST env)
* [Build / Post-deploy migration / Deploy](https://ci.backends.cms.gov/non-oc/job/Blue%20Button/job/Z%20AUTO%20-%20Build%20Deploy%20Migrate%20App%20AMI/14/console) (TEST env - using the AMI built from the Pre-deploy migrations job)

**Review by JFU-GIT — APPROVED** [review]
Looks good

**Review by njdister — CHANGES_REQUESTED** [review]
As we discussed, you'll need to stand up a canary instance with the new release AMI as part of the process to run pre-deploy migrations in the expected way.

File: Jenkinsfiles/Jenkinsfile.cbc-auto-build-deploy-migrate-app-ami [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+    stage("Determine Ansible Config: Pre-deploy") {
+      // TODO: Thinking this will need to be run separately for Pre/Post migrations since instances change post deploy
+      when {
+        expression { params.APP_MIGRATIONS == "Pre-deploy" }
+      }
+      steps {
+        script {
+          echo("Setting ansible config for tagged deployment instance")
+          ANSIBLE_INVENTORY = "inventory/ec2.py"
+          ANSIBLE_SUBSET = "tag_Function_app_AppServer"
```
Comment (njdister) [inline Jenkinsfiles/Jenkinsfile.cbc-auto-build-deploy-migrate-app-ami]: It doesn't look like these parameters change for the entire duration of the job wether they are pre or post deploy, so I'd guess you can set them once at the top of the job in the environment vars section and be done with it.

**Comment by oragame** [comment]
> As we discussed, you'll need to stand up a canary instance with the new release AMI as part of the process to run pre-deploy migrations in the expected way.

@njdister Thanks for the good feedback! Addressed the Pre-deploy migrations by adding a Canary stage. And some more refactoring and cleanups as well. Ready to review again.

**Review by njdister — APPROVED** [review]
As we discussed it would be good to reconcile migration consistency between pre and post deployment (having developers choose inlined vs running external job migrations) but from an operation standpoint this is good to go. 🚀

**Comment by oragame** [comment]
> As we discussed it would be good to reconcile migration consistency between pre and post deployment (having developers choose inlined vs running external job migrations) but from an operation standpoint this is good to go. 🚀

@njdister - went ahead and took care of this, you can check it out [here](https://ci.backends.cms.gov/non-oc/job/Blue%20Button/job/Z%20AUTO%20-%20Build%20Deploy%20Migrate%20App%20AMI/31/console).
I am adding the upper env options, and merging this now! thanks!

**Pull request merged**